pkg/etcd-backup-operator/controller/s3backup.go

```diff
@@ -0,0 +1,235 @@
+		awsSecret:   s3.AWSSecret,
+		kubecli:     b.kubecli,
+	}
+	return s3b.saveSnap()
+}
+
+// ToS3Prefix concatenates s3Prefix, S3V1, namespace, clusterName to a single s3 prefix.
+// the concatenated prefix determines the location of S3 backup files.
+func ToS3Prefix(s3Prefix, namespace, clusterName string) string {
+	return path.Join(s3Prefix, S3V1, namespace, clusterName)
+}
+
+type s3Backup struct {
+	namespace   string
+	name        string
+	clusterName string
+
+	s3bucket  string
+	prefix    string
+	awsSecret string
+
+	kubecli kubernetes.Interface
+}
+
+func (s *s3Backup) saveSnap() error {
+	podList, err := s.kubecli.Core().Pods(s.namespace).List(k8sutil.ClusterListOpt(s.clusterName))
+	if err != nil {
+		return err
+	}
+	var pods []*v1.Pod
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		if pod.Status.Phase == v1.PodRunning {
+			pods = append(pods, pod)
+		}
+	}
+
+	if len(pods) == 0 {
+		msg := "no running etcd pods found"
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+	member, rev := getMemberWithMaxRev(pods, nil)
+	if member == nil {
+		msg := "no reachable member"
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+
+	logrus.Infof("saving backup for cluster (%s) ...", s.clusterName)
+	if err := s.writeSnap(member, rev); err != nil {
+		msg := fmt.Sprintf("saving backup for cluster (%s) failed: %v", s.clusterName, err)
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+	logrus.Infof("saving backup for cluster (%s) succeeded", s.clusterName)
+	return nil
+}
+
+func (s *s3Backup) writeSnap(m *etcdutil.Member, rev int64) error {
+	cfg := clientv3.Config{
+		Endpoints:   []string{m.ClientURL()},
+		DialTimeout: constants.DefaultDialTimeout,
+		TLS:         nil,
+	}
+	etcdcli, err := clientv3.New(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create etcd client (%v)", err)
+	}
+	defer etcdcli.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	resp, err := etcdcli.Maintenance.Status(ctx, m.ClientURL())
+	cancel()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
+	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+	defer rc.Close()
+	if err != nil {
+		return fmt.Errorf("failed to receive snapshot (%v)", err)
+	}
+	return s.saveToS3(resp.Version, rev, rc)
+}
+
+// NewS3Uploader returns a Uploader that is configured with proper aws config and credentials derived from Kubernetes secret.
+func newS3Uploader(bucket, prefix string, secret *v1.Secret) (*s3manager.Uploader, error) {
+	so, err := newAWSConfig(secret, prefix)
+	if err != nil {
+		return nil, err
+	}
+	return newFromSessionOpt(bucket, prefix, *so)
+}
+
+func newAWSConfig(secret *v1.Secret, prefix string) (*session.Options, error) {
+	dir := filepath.Join(operatorRoot, "aws", prefix)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return nil, err
```

Comment:
Probably wrap most errors like you've done on line 152 for better context:
`return nil, fmt.Errorf("setup AWS config failed: mkdir (%v) failed: %v", dir, err)`
Can you also double check anywhere else for the same.